docs/create_your_own_brick.rst

```diff
@@ -0,0 +1,218 @@
+Create your own brick
+=====================
+
+This tutorial explains how to create a custom brick, which is useful if you
+want to factorize a specific sequence of operations (which can be made of
+bricks themselves) into a single operation so that you can easily reuse it.
+
+The first part of this tutorial lists the requirements and optional components
+that a brick should/can implement while the second part describes the
+construction step by step of a simple toy brick.
+
+This tutorial assumes that you are already familiar with
+:doc:`bricks <bricks_overview>`.
+
+
+Bricks ingredients and recipe
+-----------------------------
+
+All the bricks in blocks inherit directly or indirectly from the
+:class:`.Brick`. However there is already a rich inheritance hierarchy of
+bricks implemented in blocks and thus, you should consider which brick level
+you wish to inherit from. Bear it mind that multiple inheritance is often
+possible and advocated whenever it makes sense.
+
+Here are examples of possible bricks to inherit from:
+
+* :class:`.Sequence`: a sequence of bricks.
+* :class:`.Initializable`: a brick that defines a same initialiation scheme
+  (weights and biases) for all its children.
+* :class:`.Feedforward`: declares an interface for bricks with one input and
+  one output.
+* :class:`.Linear`: a linear transformation with optional bias. Inherits from
+  :class:`.Initializable` and :class:`.Feedforward`.
+* many mores!
+
+Let's say that you want to create a brick from scracth, simply inheriting
+from :class:`.Brick`, then you should consider overwriting the following
+methods (strictly speaking, all these methods are optional):
+
+* :meth:`.Brick.__init__`: you should pass by argument the attributes of your
+  bricks. It is also in this method that you should create the potential
+  "children bricks" that belongs to your brick. The initialiazation of the
+  attributes can be lazy as described in a further paragraph.
+* :meth:`you_decide_which_name`: you need to implement a method that actually
+  implements the operation of the brick, taking as arguments the inputs
+  of the brick and returning its outputs. It can have any name and for simple
+  bricks is often named ``apply``. You can decorate it with the
+  :func:`.application` decorator, as explained in the next section.
+* :meth:`.Brick._allocate`: you should implement this method if your brick
+  needs to allocate its parameters.
+* :meth:`.Brick._initialize`: you should implement this method if you need to
+  initialize parameters of your brick.
+* :meth:`.Brick._push_allocation_config`: you should consider overwriting
+  this method if you want to allocate the children bricks in a specific way.
+* :meth:`.Brick._push_initialization_config`: you should consider method if
+  you want to initialize the children bricks in a specific way.
```

Comment:
It worth mentioning here that usually one should just inherit from `Initializable` and it is going to initialize children recursively.